omniledger/service/service.go

```diff
@@ -191,6 +195,44 @@ func (s *Service) AddTransaction(req *AddTxRequest) (*AddTxResponse, error) {
 
 	s.txBuffer.add(string(req.SkipchainID), req.Transaction)
 
+	if req.InclusionWait != nil {
+		// Wait for InclusionWait new blocks and look if our transaction is in it.
+		current := s.state.getLast(req.SkipchainID)
+		interval, err := LoadBlockIntervalFromColl(s.GetCollectionView(req.SkipchainID))
+		if err != nil {
+			return nil, errors.New("couldn't get collectionView: " + err.Error())
+		}
+		ctHash := req.Transaction.Instructions.Hash()
+		found := false
+	inclusionWait:
+		for *req.InclusionWait > 0 {
+			time.Sleep(interval)
+			next := s.state.getLast(req.SkipchainID)
+			if !next.Equal(current) {
+				log.Lvl3("Found new block")
+				_, bodyI, err := network.Unmarshal(s.db().GetByID(next).Payload, cothority.Suite)
+				body, ok := bodyI.(*DataBody)
+				if err != nil || !ok {
+					return nil, fmt.Errorf("couldn't unmarshal body: %s - type was %t", err, ok)
+				}
+				for _, ct := range body.Transactions {
+					log.Lvlf3("Searching transaction %x == %x", ct.Instructions.Hash(), ctHash)
+					if bytes.Compare(ct.Instructions.Hash(), ctHash) == 0 {
+						log.Lvl3("found")
+						found = true
+						break inclusionWait
+					}
+				}
+				current = next
+			} else {
+				log.Lvl3("no new block")
+			}
+			*req.InclusionWait--
+		}
+		if !found {
+			return nil, errors.New("didn't find transaction in blocks")
+		}
```

Comment:
While this implementation works fine, we might be unmarshalling a lot of data, especially there are concurrent transactions. I feel it's more elegant to just maintain a map of channels that signals inclusion. Further, the client will know immediately when their block is included. Concretely, I'm imagining something like:
- on AddTransaction:
  - if InclusionWait > 0, create a channel and put it in an InclusionWait map 
  - wait on this channel
- on updateCollection:
  - we're unmarshalling `DataBody` anyway, so at that point we can check if any transaction are in the map
  - if we found one, then write to the channel
- on AddTransaction:
  - if a signal comes back, we know our transaction is included
  - if InclusionWait times out, return an error (the InclusionWait is in real-time instead of number of block intervals)

Reply:
Done.